### PR TITLE
Adds a codespell linter

### LINKED ISSUE #1536: Adding Codespell
It would be cool to add codespell to our CI. Let me know if you want this done. (Also its a fairly simple python linter, so we could add it to ruff itself eventually, but that is probably out of scope for now).

### PR DESCRIPTION
**Pull request by colin99d**
Fixes #1536

### CODE REVIEW

**Comment by not-my-profile** [comment]
The log says

> Running codespell on '' with the following options...

so is it actually doing anything?

When I run codespell locally on the `src` directory it reports a bunch of useless `crate ==> create` errors.

**Comment by colin99d** [comment]
Good catch. I thought all of the crate issues were in the `./target` folder, which isnt in the repo. But it looks like there are a couple in other places. Will fix now.

**Comment by colin99d** [comment]
Linter is working now! Its currently failing on a couple errors (showing the value). Once @charliermarsh confirms he wants it I will resolve those errors and get this thing merged.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -181,3 +181,25 @@ jobs:
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.9"
+          architecture: x64
+
+      - uses: actions/cache@v3
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-linting-${{ hashFiles('**/poetry.lock') }}
+          restore-keys: ${{ runner.os }}-linting-${{ hashFiles('**/poetry.lock') }}
```
Comment (andersk) [inline .github/workflows/ci.yaml:202]: This action doesn’t interact with `poetry.lock`.

**Review by andersk — COMMENTED** [review]
There’s an official GitHub Action from the Codespell project that might make more sense: https://github.com/codespell-project/actions-codespell

**Comment by charliermarsh** [comment]
@colin99d - I def appreciate this, but I'd like to keep the development tooling within the Rust ecosystem where possible. Would you mind modifying this to use the [`typos`](https://github.com/crate-ci/typos) crate instead?

**Comment by colin99d** [comment]
@charliermarsh I definitely agree, I just switched it and some more errors were found. Let me fix the spelling mistakes and then we should be good to go.

**Comment by charliermarsh** [comment]
Thanks!

File: _typos.toml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[default.extend-words]
+froms = "froms"
+fo = "fo"
```
Comment (charliermarsh) [inline _typos.toml:3]: Should we just change those to some valid word, to avoid false negatives?

Comment (colin99d) [reply _typos.toml:3]: Im on it.

Comment (charliermarsh) [reply _typos.toml:3]: Thanks tons

Comment (colin99d) [reply _typos.toml]: Np. Done.

**Pull request merged**